File: R/gmm.R
## gmm.R - Gaussian Mixture Models
##
## by Artem Sokolov

## For all functions:
##
## (N) - Number of samples
## (K) - Number of components
## (p) - Dimensionality
##
## Gaussian mixture model parameters gmm must contain:
##	mu    - [K x p] matrix of component means
##	sigma - [K x p x p] array of covariance matrices for each component
##	pi    - [K x 1] vector of mixture coefficients (has to sum up to 1)


## Generates a gmm instance
##
## Inputs:
##	K - number of components
##	p - dimensionality
##	mu, sigma - general orientation of the model
##
## Output:
##	the model
GMM.make <- function( K, p, mu=rep(0, p), sigma=diag(p) )
  {
    gmm <- list()
    gmm$pi <- rep( 1/K, K )
    gmm$mu <- rmvnorm( K, mu, sigma )
    gmm$sigma <- array( 0, c(K,p,p) )
    for( i in 1:K )
      gmm$sigma[i,,] <- sigma
    gmm
  }

## Generates data from a provided GMM
##
## Inputs:
##	N - number of samples
##	gmm - the model (with dimensionality p)
##
## Output:
##	[N x p] data matrix
GMM.genData <- function( N, gmm )
  {
    K <- nrow( gmm$mu )
    p <- ncol( gmm$mu )
    x <- matrix( 0, N, p )
    j <- sample( 1:K, N, TRUE, gmm$pi )

    for( i in 1:K )
      {
        k <- which( j == i )
        n <- length( k )
        if( n > 0 )
          x[ k, ] <- rmvnorm( n, gmm$mu[i,], gmm$sigma[i,,] )
      }

    x
  }

## Computes component responsibilities and posterior probabilities
##
## Inputs:
## 	x - [N x p] data matrix
## 	gmm - the model
##
## Outputs:
##	resp - [N x K] matrix of responsibilities
##	LL   - [N x 1] vector of log likelihoods
GMM.resp <- function( x, gmm )
  {
    if( is.list(x) == TRUE )
      stop( "GMM.resp: x must be a single observation sequence" )
    
    x <- as.matrix(x)
    gmm$pi <- gmm$pi/sum(gmm$pi)

    N <- nrow(x)
    p <- ncol(x)
    K <- length(gmm$pi)

    if( SM.getOption( "pure.R" ) == FALSE )
      {
        res <- .C( "GMMrespR", as.numeric(x), as.numeric(gmm$mu), as.numeric(gmm$sigma), as.numeric(gmm$pi),
                  as.integer(N), as.integer(p), as.integer(K), resp=numeric(N*K), LL=numeric(N),
                  status=integer(1), info=integer(N), DUP=FALSE )

        if( res$status == 1 )
          stop( paste("GMM.resp: Covariance matrix is singular for component", res$info[1]) )
        
        else
          return( list(
                       resp=matrix(res$resp, N, K),
                       LL=matrix(res$LL, N, 1),
                       outliers=which(res$info == 1)
                       ) )
      }
    else
      {
        ## Compute posterior probabilties using Bayes rule
        post.prob <- matrix( 0, N, K )
        for( i in 1:K )
          post.prob[,i] <- tryCatch(gmm$pi[i] * dmvnorm(x, gmm$mu[i,,drop=FALSE], as.matrix(gmm$sigma[i,,])),
                                    error=function(e)
                                    stop(paste("GMM.resp: Covariance matrix is singular for component", i))
                                    )

        ## Compute responsibilities by scaling posteriors
        p.x <- apply( post.prob, 1, sum )
        resp <- apply( post.prob, 2, "/", p.x )

        ## Return responsibilities and log likelihood of the posteriors
        return( list( resp=as.matrix(resp), LL=as.matrix(log(p.x)), outliers=which(p.x==0) ) )
      }
  }

## Computes log-likelihood of new observation sequences given a previously trained model
##
## Inputs:
##	x - Either a single observation sequence matrix or a list of multiple observation sequences
##	gmm - the model
##
## Output:
##	If x is a single observation sequence, returns log-likelihood of that sequence
##	If x is a list of multiple observation sequences, returns a list with each element being log-likelihood
##		of the corresponding sequence in x
GMM.ll <- function( x, gmm )
  {
    if( is.list( x ) == TRUE )
      lapply( x, sys.function(), gmm )
    else
      sum( GMM.resp( x, gmm )$LL )
  }

## Expectation Maximization for multiple observation sequences
##
## (Ni) - Number of samples in training sequence i
## (Mi) - Number of samples in validation sequence i
##
## Inputs:
## xL - list of [Ni x p] data matrices
## K - Number of components
## vL - list of [Mi x p] validation data matrices
## gmm.init - optional initialization point (can be partially specified)
## cov.reg - covariance matrix regularization (towards identity) parameter, must be in [0,1]
##
## Stopping Criteria:
## tol		- relative tolerance for convergence
## LLstop	- stop iterating when LL exceeds this
## max.iter	- cap on the number of iterations
##
## Outputs:
## gmm - Gaussian Mixture model
GMM.learn <- function( xL, K, vL=NULL, gmm.init=NULL, cov.reg=0.0, tol=1e-03, LLstop=Inf, min.iter=3, max.iter=Inf )
  {
    ## Argument type verification
    if( is.list(xL) == FALSE )
      xL <- list( xL )
    if( is.null(vL) == FALSE && is.list(vL) == FALSE )
      vL <- list( vL )

    ## Further argument verification
    if( cov.reg < 0 ) cov.reg <- 0
    if( cov.reg > 1 ) cov.reg <- 1
    
    nseq <- length(xL)
    nval <- length(vL)

    d.cat( 1, "Learning a Gaussian Mixture Model with", K,"component(s) for", nseq, "observation sequences\n" )
    
    ## Concatenate all training sequences together
    x <- NULL
    for( i in 1:nseq )
      x <- rbind( x, as.matrix(xL[[i]]) )
    N <- nrow(x)

    ## Concatenate all validation sequences together
    v <- NULL
    for( i in seq( 1, nval, length=nval ) )
      v <- rbind( v, as.matrix(vL[[i]]) )
    M <- nrow(v)
        
    p <- ncol( x )

    ## Special case
    if( K == 1 )
      return( list( pi=1, mu = matrix( apply( x, 2, mean ), nrow=1 ), sigma = array( cov(x), c(1,p,p) ) ) )

    ## Initialize GMM
    gmm <- GMM.make( K, p, apply( x, 2, mean ), cov(x) )
    if( is.null( gmm.init$mu ) == FALSE )	gmm$mu <- gmm.init$mu
    if( is.null( gmm.init$sigma ) == FALSE )	gmm$sigma <- gmm.init$sigma
    if( is.null( gmm.init$pi ) == FALSE )	gmm$pi <- gmm.init$pi

    ########################
    ## Attempt to run C code
    if( SM.getOption("pure.R") == FALSE )
      {
        ## There is a weird issue that occurs when C code is the first thing ran after R startup
        ## Apparently random number generation using rnorm() in Rmath.h always returns the same number
        ## The problem seems to be alleviated when the first call to the function is made from R
        ## For this purpose:
        rnorm( 1, 0, 1 )

        ## Attempt to run C code
        if( is.finite( max.iter ) == FALSE ) max.iter <- -1
        res <- .C( "GMMlearnR",
                  as.numeric(x), as.numeric(v), as.integer(N), as.integer(M), as.integer(K), as.integer(p),
                  as.numeric(gmm$mu), as.numeric(gmm$sigma), as.numeric(gmm$pi),
                  as.numeric( cov.reg ),
                  as.numeric(tol), as.numeric(LLstop),
                  as.integer(min.iter), as.integer(max.iter),
                  as.integer(SM.getOption("debug.output")),
                  pi = numeric( K ), mu = numeric( K*p ), sigma = numeric( K*p*p ), status = integer( 1 ),
                  NAOK=TRUE, DUP=FALSE )

        if( res$status == 1 )
          stop( "Covariance matrix is singular for one of the clusters\n" )
        else if( res$status == 2 )
          stop( "Some points were not assigned to a cluster during the E-Step\n" )
        else if( res$status == 3 )
          stop( "No points were assigned to one of the clusters\n" )
        else
          return( list(
                       pi = res$pi,
                       mu = matrix( res$mu, K, p ),
                       sigma = array( res$sigma, c( K, p, p ) )
                       ) )
      }

    #################################
    ## Revert to the R Implementation
    d.cat( 1, "Currently running R version of the GMM.learn code\n" );

    gmm.prev <- list()

    LLval <- 0
    LLval.prev <- 0
    LLbase <- 0
    LLprev <- 0
    iter <- 0
    while( TRUE )
      {
        ## Validation log-likelihood stopping criterion
        if( nval > 0 )
          {
            LLval <- sum( GMM.resp( v, gmm )$LL )
            d.cat( 3, "Validation log-likelihood: ", LLval, "\n" )
            if( iter <= min.iter || LLval >= LLval.prev )
              {
                LLval.prev <- LLval
                gmm.prev <- gmm
              }
            else
              {
                d.cat( 1, "Final log likelihod of training data after", iter-1, "iterations:", LLprev, "\n" )
                d.cat( 1, "Validation log likelihood:", LLval.prev, "\n" )
                return( gmm.prev )
              }
          }
            
        ##
        ## E STEP - compute responsibilities
        ##
        pp <- GMM.resp( x, gmm )

        if( length( pp$outliers ) > 0 )
          stop( "Some points were not assigned to a cluster during the E-Step\n" )
        
        resp <- pp$resp
        LL <- sum( pp$LL )

        d.cat( if( (iter%%10)==0 ) 2 else 3,
              "Log likelihood at iteration", iter, ":", LL, "\n" )
        
        ## Verify stopping criteria
        if( iter >= max.iter ) break
        if( LL > LLstop ) break
        if( is.infinite(LL) ) break
        if( LL == LLprev ) break
        
        if( iter <= (min.iter-1) ) LLbase <- LL
        else if( (LL-LLbase) < (1+tol)*(LLprev-LLbase) ) break

        ##
        ## M STEP - Compute the new parameters
        ##
        for( i in 1:K )
          {
            gamma.i <- matrix( resp[,i], N, p )
            norm.fac <- sum( resp[,i] )

            if( norm.fac == 0 )
              stop( paste( "No points were assigned to cluster", i, "during the E-Step\n" ) )
            
            gmm$mu[i,] <- apply( (gamma.i*x), 2, sum ) / norm.fac
            x.shifted <- t( t( x ) - gmm$mu[i,] )
            gmm$sigma[i,,] <- (t( x.shifted * gamma.i ) %*% x.shifted) / norm.fac

            if( cov.reg > 0 )
              gmm$sigma[i,,] <- cov.reg * diag(p) + (1-cov.reg) * gmm$sigma[i,,]
            
            gmm$pi[i] <- norm.fac / N
          }
        gmm$pi <- gmm$pi / sum(gmm$pi)

        LLprev <- LL
        iter <- iter+1
      }

    d.cat( 1, "Final log likelihod after", iter, "iterations:", LL, "\n" )
    gmm
  }

File: src/gmm.c
// gmm.c - Gaussian Mixture Model routines
//
// by Artem Sokolov

#include <Rmath.h>

#include "matrix.h"
#include "stats.h"

// Computes cluster responsibilities and posterior log-likelihood
//
// N - number of samples
// p - observation dimensionality
// K - number of components in the model
//
// Inputs:
//      x - [N x p] matrix of observations
//	mu - [K x p] matrix of component means
//	sigma - [K x p x p] array of covariance matrices for each component
//	pi - [K x 1] vector of mixture coefficients (has to sum up to 1)
//
// Outputs:
//	resp - [N x K] matrix of responsibilities
//	LL   - [N x 1] vector of log likelihoods
//
// Returns:
//	0 - everything is OK
//	1 - covariance matrix for cluster info[0] is singular
//	2 - points which( info == TRUE ) were not assigned to any cluster
int GMM_resp( double** x, double** mu, double*** sigma, double* pi,
	      int N, int p, int K,
	      double** resp, double* LL, int* info )
{
  // Create local storage
  int i, j;
  int status;
  double* p_x = make1D( N );
  double** post_prob = make2D( K, N );

  // Normalize pi
  mult_const_vec( K, 1.0 / vec_sum( K, pi ), pi );

  // Compute posterior probabilities using Bayes rule
  for( i = 0; i < K; i++ )
    {
      status = dmvnorm( N, p, x, mu[i], sigma[i], post_prob[i] );

      // Check for numerical stability
      if( status > 0 )
	{
	  free( p_x );
	  free2D( post_prob, K );
	  info[0] = i+1;
	  return status;
	}

      mult_const_vec( N, pi[i], post_prob[i] );
    }

  // Compute responsibilities by scaling posteriors
  for( i = 0; i < N; i++ ) p_x[i] = 0.0;
  for( i = 0; i < K; i++ )
    for( j = 0; j < N; j++ )
      p_x[j] += post_prob[i][j];

  for( i = 0; i < N; i++ )
    {
      if( p_x[i] == 0 )
	{
	  status = 2;
	  info[i] = 1;
	  for( j = 0; j < K; j++ )
	    resp[i][j] = 0.0;
	}
      else
	{
	  info[i] = 0;
	  for( j = 0; j < K; j++ )
	    resp[i][j] = post_prob[j][i] / p_x[i];
	}
    }

  // Compute log likelihood values
  for( i = 0; i < N; i++ )
    LL[i] = log( p_x[i] );

  // Free up local memory
  free( p_x );
  free2D( post_prob, K );
  return status;
}


// Expectation Maximization for multiple observation sequences
//
// Inputs:
// 	x - [N x p] matrix of training samples
// 	v - [M x p] matrix of validation samples 
// 	N - number of training samples
// 	M - number of validation samples
// 	K - Desired number of components
// 	p - Observation dimensionality
//
// Control variables:
//	tol - relative tolerance for convergence of training log-likelihood
//	LLstop - iterations will stop when training log-likelihood exceeds this value
//	max_iter - maximum number of iterations to perform (use negative number to denote infinity)
//	debug_output - amount of output produced by the algorithm
//		0 - none
//		1 - basic
//		2 - verbose
//
// Outputs:
//	pi - [K x 1] vector of mixture coefficients
// 	mu - [K x p] matrix of component means
//	sigma - [K x p x p] matrix of component covariance matrices
//
// Returns:
//	0 - Everything is OK
//	1 - Covariance matrix is singular for one of the clusters
//	2 - Some points were not assigned to a cluster during the E-step
//	3 - No points assigned to one of the clusters
int GMM_learn( double** x, double** v, int N, int M, int K, int p,
	       double* pi, double** mu, double*** sigma,
	       double cov_reg, double tol, double LLstop, 
	       int min_iter, int max_iter, int debug_output )
{
  if( debug_output > 0 )
    printf( "Currently running C version of the GMM.learn code\n" );

  // Create local storage
  int status = 0;
  int i, j, k, l;
  int iter;
  double LL, LLprev, LLbase;
  double LLval, LLval_prev;

  int* info = malloc( N*sizeof(int) );
  double* LLvec = make1D( N );
  double** resp = make2D( N, K );
  double** x_shifted = make2D( N, p );

  double* v_LLvec = NULL;
  double** v_resp = NULL;
  double* pi_prev = NULL;
  double** mu_prev = NULL;
  double*** sigma_prev = NULL;

  if( M > 0 )
    {
      v_LLvec = make1D( M );
      v_resp = make2D( M, K );
      pi_prev = make1D( K );
      mu_prev = make2D( K, p );
      sigma_prev = make3D( K, p, p );
    }

  // Run EM algorithm
  iter = 0;
  while( 1 )
    {
      // Validation log-likelihood stopping criterion
      if( M > 0 )
	{
	  GMM_resp( v, mu, sigma, pi, M, p, K, v_resp, v_LLvec, info );
	  LLval = vec_sum( M, v_LLvec );
	  if( debug_output > 2 )
	    printf( "Validation log-likelihood: %f\n", LLval_prev );

	  if( iter <= min_iter || LLval >= LLval_prev )
	    {
	      LLval_prev = LLval;
	      cp_vec_vec( K, pi, pi_prev );
	      cp_mat_mat( K, p, mu, mu_prev );
	      cp_3D_3D( K, p, p, sigma, sigma_prev );
	    }
	  else
	    {
	      if( debug_output > 0 )
		{
		  printf( "Final log likelihood of training data after %d iterations is %f\n", iter-1, LLprev );
		  printf( "Validation log likelihood: %f\n", LLval_prev );
		}
	      cp_vec_vec( K, pi_prev, pi );
	      cp_mat_mat( K, p, mu_prev, mu );
	      cp_3D_3D( K, p, p, sigma_prev, sigma );

	      status = 0;
	      break;
	    }
	}
      

      //
      // E Step - compute responsibilities
      //
      status = GMM_resp( x, mu, sigma, pi, N, p, K, resp, LLvec, info );
      if( status != 0 ) break;

      LL = vec_sum( N, LLvec );
      if( isnan(LL) || isinf(LL) )
	{
	  status = 1;
	  break;
	}
	
      if( debug_output > 2 || (debug_output > 1 && (iter%10) == 0) )
	printf( "Log likelihood at iteration %d is %f\n", iter, LL );

      // Verify stopping criteria
      if( K == 1 && iter == 1 ) break;
      if( max_iter > 0 && iter >= max_iter ) break;
      if( LL > LLstop ) break;

      if( iter <= (min_iter-1) ) LLbase = LL;
      else if( (LL-LLbase) < (1+tol)*(LLprev-LLbase) ) break;

      // 
      // M Step - compute the new parameters
      //
      for( k = 0; k < K; k++ )
	{
	  pi[k] = 0.0;
	  for( i = 0; i < N; i++ )
	    pi[k] += resp[i][k];

	  if( pi[k] == 0 )
	    {
	      status = 3;
	      break;
	    }

	  for( j = 0; j < p; j++ )
	    {
	      mu[k][j] = 0.0;
	      for( i = 0; i < N; i++ )
		mu[k][j] += x[i][j] * resp[i][k];
	      mu[k][j] /= pi[k];
	    }

	  for( i = 0; i < N; i++ )
	    for( j = 0; j < p; j++ )
	      x_shifted[i][j] = x[i][j] - mu[k][j];

	  for( i = 0; i < p; i++ )
	    for( j = 0; j < p; j++ )
	      {
		sigma[k][i][j] = 0.0;
		for( l = 0; l < N; l++ )
		  sigma[k][i][j] += resp[l][k] * x_shifted[l][i] * x_shifted[l][j];
		sigma[k][i][j] /= pi[k];
	      }

	  if( cov_reg > 0 )
	    {
	      for( i = 0; i < p; i++ )
		for( j = 0; j < p; j++ )
		  {
		    sigma[k][i][j] *= (1-cov_reg);
		    if( i == j )
		      sigma[k][i][j] += cov_reg;
		  }
	    }
	}
      if( status != 0 ) break;

      mult_const_vec( K, 1.0 / vec_sum( K, pi ), pi );

      LLprev = LL;
      iter++;
    }

  // Display final log likelihood
  if( status == 0 && debug_output > 0 )
    printf( "Final log likelihood after %d iterations is %f\n", iter, LL );

  // Free up memory
  free( info );
  free( LLvec );
  free2D( resp, N );
  free2D( x_shifted, N );

  if( M > 0 )
    {
      free( v_LLvec );
      free2D( v_resp, M );
      free( pi_prev );
      free2D( mu_prev, K );
      free3D( sigma_prev, K, p );
    }

  return status;
}
